include/zephyr/sys/time_units.h

```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
-		} else {
-			/* Slower algorithm but input is first divided before being multiplied
-			 * which prevents overflow of intermediate value.
-			 */
-			return (t / from_hz) * to_hz + ((t % from_hz) * to_hz + off) / from_hz;
-		}
-	}
-}
+
+/* true if the conversion is the identity */
+#define z_tmcvt_identity(__from_hz, __to_hz) \
+	((__to_hz) == (__from_hz))
+
+/* true if the conversion requires a simple integer multiply */
+#define z_tmcvt_mul_ratio(__from_hz, __to_hz) \
+	((__to_hz) > (__from_hz) && (__to_hz) % (__from_hz) == 0U)
+
+/* true if the conversion requires a simple integer division */
+#define z_tmcvt_div_ratio(__from_hz, __to_hz) \
+	((__from_hz) > (__to_hz) && (uint64_t) (__from_hz) % (__to_hz) == 0U)
+
+/*
+ * Compute the offset needed to round the result correctly when
+ * the conversion requires a simple integer division
+ */
+#define z_tmcvt_off_div(__from_hz, __to_hz, __round_up, __round_off)	\
+	((__round_off) ? ((__from_hz) / (__to_hz)) / 2 :		\
+	 (__round_up) ? ((__from_hz) / (__to_hz)) - 1 :			\
+	 0)
+
+/*
+ * Compute the offset needed to round the result correctly when
+ * the conversion requires a full mul/div
+ */
+#define z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)	\
+	((__round_off) ? (__from_hz) / 2 :				\
+	 (__round_up) ? (__from_hz) - 1 :				\
+	 0)
+
+/* General 32-bit conversion for constant frequencies */
+#define z_tmcvt_const_32(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	((uint32_t) (((uint64_t) (__t)*(__to_hz) + \
+		      z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz)))
+
+/* Fast 64-bit conversion for constant frequencies. This relies on the multiply not overflowing */
+#define z_tmcvt_const_64_fast(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(((uint64_t) (__t)*(__to_hz) + \
+	  z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz))
+
+/* Slow 64-bit conversion for constant frequencies. This avoids overflowing the multiply */
+#define z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(((uint64_t) (__t) / (__from_hz))*(__to_hz) +			\
+	 (((uint64_t) (__t) % (__from_hz))*(__to_hz) +		\
+	  z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz))
+
+/* General 64-bit conversion for constant frequencies. Uses one of the two above macros */
+#define z_tmcvt_const_64(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(Z_TMCVT_USE_FAST_ALGO(__from_hz, __to_hz) ?			\
+	 z_tmcvt_const_64_fast(__t, __from_hz, __to_hz, __round_up, __round_off) : \
+	 z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off))
+
+/* Functions for cases with non-constant frequencies */
+uint32_t _z_tmcvt_32(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+uint64_t _z_tmcvt_64(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+/* Convert, generating a 32-bit result */
```

Comment:
Agreed. *and* using `uint64_t` for `__t` instead of having the _32 macros take in 32-bit values which might lead to truncation. I'm thinking particularly of the cycle to ms conversion where cycle counts are likely to be large while ms might well be within a 32-bit range. __t is *always* used as a 64-bit value now.